Framework/PythonInterface/core/src/PythonLoggingChannel.cpp

```diff
@@ -0,0 +1,80 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+
+// local includes
+#include "MantidPythonInterface/core/PythonLoggingChannel.h"
+
+// 3rd-party includes
+#include "MantidPythonInterface/core/GlobalInterpreterLock.h"
+#include <Poco/Message.h>
+#include <boost/python/import.hpp>
+
+namespace Poco {
+
+namespace {
+// See https://docs.python.org/3/library/logging.html#logging-levels
+enum class PyLogLevel : int {
+  CRITICAL = 50,
+  ERROR = 40,
+  WARNING = 30,
+  INFO = 20,
+  DEBUG = 10,
+  NOTSET = 0,
+};
+
+PyLogLevel pythonLevel(const Message::Priority prio) {
+  switch (prio) {
+  case Message::Priority::PRIO_FATAL:
+  case Message::Priority::PRIO_CRITICAL:
+    return PyLogLevel::CRITICAL;
+  case Message::Priority::PRIO_ERROR:
+    return PyLogLevel::ERROR;
+  case Message::Priority::PRIO_WARNING:
+    return PyLogLevel::WARNING;
+  case Message::Priority::PRIO_NOTICE:
+  case Message::Priority::PRIO_INFORMATION:
+    return PyLogLevel::INFO;
+  case Message::Priority::PRIO_DEBUG:
+  case Message::Priority::PRIO_TRACE:
+    return PyLogLevel::DEBUG;
+  default:
+    return PyLogLevel::NOTSET;
+  }
+}
+
+} // namespace
+
+PythonLoggingChannel::PythonLoggingChannel() {
+  Mantid::PythonInterface::GlobalInterpreterLock gil;
+  auto logger = (boost::python::import("logging").attr("getLogger")("Mantid"));
+  m_pyLogger = std::make_unique<boost::python::object>(std::move(logger));
+}
+
+// The special behavior here is needed because Poco's LoggingFactory can be destroyed
+// after the Python interpreter was shut down.
+PythonLoggingChannel::~PythonLoggingChannel() {
+  if (Py_IsInitialized()) {
+    Mantid::PythonInterface::GlobalInterpreterLock gil;
```

Comment:
I may be reading this wrong, but I think `release` should happen if python is initialized and setting it to `nullptr` should happen if python isn't initialized.

Reply:
`m_pyLogger = nullptr;` makes `std::unique_ptr` delete the contained object which calls its destructor which in turn needs to access Python.
The `release` hack allows us to only free the memory allocated by us without calling into Python via `*m_pyLogger`'s destructor.